src/graft_dag.cpp

```diff
@@ -0,0 +1,487 @@
+// Copyright 2019-2022 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+
+#include "graft_dag.hpp"
+
+#include "gp_dag.hpp"
+#include "quartet_hybrid_request.hpp"
+#include "reindexer.hpp"
+#include "subsplit_dag.hpp"
+
+// ** Constructors
+
+GraftDAG::GraftDAG(SubsplitDAG &dag) : host_dag_(dag) {}
+
+// ** Comparators
+
+int GraftDAG::Compare(const GraftDAG &other) const {
+  return GraftDAG::Compare(*this, other);
+}
+
+int GraftDAG::Compare(const GraftDAG &lhs, const GraftDAG &rhs) {
+  const SubsplitDAG &host_a = lhs.GetHostDAG();
+  const SubsplitDAG &host_b = rhs.GetHostDAG();
+  // (1) Compare host DAGs.
+  int host_diff = SubsplitDAG::Compare(host_a, host_b);
+  if (host_diff != 0) {
+    return host_diff;
+  }
+  // (2) Compare graft nodes.
+  BitsetVector nodes_a = lhs.GetSortedVectorOfNodeBitsets();
+  BitsetVector nodes_b = rhs.GetSortedVectorOfNodeBitsets();
+  if (nodes_a != nodes_b) {
+    return (nodes_a < nodes_b) ? -1 : 1;
+  }
+  // (3) Compare graft edges.
+  BitsetVector edges_a = lhs.GetSortedVectorOfEdgeBitsets();
+  BitsetVector edges_b = rhs.GetSortedVectorOfEdgeBitsets();
+  if (edges_a != edges_b) {
+    return (edges_a < edges_b) ? -1 : 1;
+  }
+  return 0;
+}
+
+int GraftDAG::CompareToDAG(const SubsplitDAG &other) const {
+  return GraftDAG::CompareToDAG(*this, other);
+}
+
+int GraftDAG::CompareToDAG(const GraftDAG &lhs, const SubsplitDAG &rhs) {
+  // Compare taxon counts.
+  int taxon_diff = lhs.TaxonCount() - rhs.TaxonCount();
+  if (taxon_diff != 0) {
+    return taxon_diff;
+  }
+  // Compare nodes.
+  BitsetVector nodes_a = lhs.GetSortedVectorOfNodeBitsets();
+  BitsetVector nodes_b = rhs.GetSortedVectorOfNodeBitsets();
+  if (nodes_a != nodes_b) {
+    return (nodes_a < nodes_b) ? -1 : 1;
+  }
+  // Compare edges.
+  BitsetVector edges_a = lhs.GetSortedVectorOfEdgeBitsets();
+  BitsetVector edges_b = rhs.GetSortedVectorOfEdgeBitsets();
+  if (edges_a != edges_b) {
+    return (edges_a < edges_b) ? -1 : 1;
+  }
+  return 0;
+}
+
+// ** Modify GraftDAG
+
+void GraftDAG::CreateGraftNode(const Bitset &node_subsplit) {
+  // Do not add node to DAG if already exists in DAG.
+  bool node_exists_in_dag = host_dag_.ContainsNode(node_subsplit);
```

Comment:
Seems like we could have a nice `ContainsNode` method.

Reply:
Fixed.